core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -170,6 +170,13 @@ class KafkaApis(val requestChannel: RequestChannel,
     val response =
       if (authorize(request.session, ClusterAction, Resource.ClusterResource)) {
         val (result, error) = replicaManager.stopReplicas(stopReplicaRequest)
+        result.map{case (topicPartition, errorCode) => {
```

Comment:
Hey @becketqin, it's an interesting point. It does seem like `GroupMetadataManager.addGroup` is not safe with a concurrent call to `GroupMetadataManager.removeGroupsForPartition`. Perhaps it would be sufficient to grab the partition lock in `GroupMetadataManager` and verify that we own the partition at the time the group is added? That would ensure that the group cache is always consistent with the state of the owned partitions, which seems like the main gap at the moment. It could still happen that a partition is moved with a rebalance in progress, but I think we already handle that case.

In general, it is possible for two consumers to believe that they are assigned the same partition at the same time, but as you said, there should be only one member which can ever commit offsets.